REGRESSION(296360@main) Broke WKSeparatedImageViews

etiennesegonzac opened the pull request:
#### 9efbae20a35b7d62ce977f84d8dc9df0c4d796c0
<pre>
REGRESSION(296360@main) Broke WKSeparatedImageViews
<a href="https://bugs.webkit.org/show_bug.cgi?id=294851">https://bugs.webkit.org/show_bug.cgi?id=294851</a>
&lt;<a href="https://rdar.apple.com/149583665">rdar://149583665</a>&gt;

Reviewed by Matt Woodrow.

`RemoteLayerBackingStoreProperties::updateCachedBuffers` shouldn&apos;t consume
the `m_bufferHandle` for WKSeparatedImageViews.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
Pass `hostingView` to `updateCachedBuffers` too.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStoreProperties::applyBackingStoreToLayer):
Always check `m_bufferHandle` before using it.
(WebKit::RemoteLayerBackingStoreProperties::updateCachedBuffers):
Don&apos;t consume the buffer handle for WKSeparatedImageViews.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
Pass `hostingView` to `updateCachedBuffers` too.

Canonical link: <a href="https://commits.webkit.org/296536@main">https://commits.webkit.org/296536@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6254e86b24e08ad250b23a6fa5f0fb3d0e8a0cae

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108828 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28488 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18913 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114036 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/59187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29172 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37052 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82688 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/59187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111776 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23183 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98021 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63127 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22606 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58741 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92555 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16206 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117157 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35878 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/26501 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91703 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36251 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94287 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91513 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23301 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36412 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14168 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31750 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35778 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41310 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35480 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38825 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37165 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
```diff
@@ -671,7 +674,11 @@ static bool hasValue(const ImageBufferBackendHandle& backendHandle)
     }
 
+#if HAVE(CORE_ANIMATION_SEPARATED_LAYERS)
+    if (!m_contentsBuffer && (!hostingView || ![hostingView isKindOfClass:[WKSeparatedImageView class]])) {
```
These checks for `WKSeparatedImageView` should be at the top of this function, as part of the early exit.

etiennesegonzac committed:
REGRESSION(296360@main) Broke WKSeparatedImageViews
https://bugs.webkit.org/show_bug.cgi?id=294851
<rdar://149583665>

Reviewed by Matt Woodrow.

`RemoteLayerBackingStoreProperties::updateCachedBuffers` shouldn't consume
the `m_bufferHandle` for WKSeparatedImageViews.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
Pass `hostingView` to `updateCachedBuffers` too.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStoreProperties::applyBackingStoreToLayer):
Always check `m_bufferHandle` before using it.
(WebKit::RemoteLayerBackingStoreProperties::updateCachedBuffers):
Don't consume the buffer handle for WKSeparatedImageViews.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
Pass `hostingView` to `updateCachedBuffers` too.

Canonical link: https://commits.webkit.org/296536@main
